### ISSUE TITLE
LDAP user could not accept nor reject group folder share

### ISSUE THREAD

**Issue by g1smo** [nextcloud/server#21896]
<!--
Thanks for reporting issues back to Nextcloud!

Note: This is the **issue tracker of Nextcloud**, please do NOT use this to get answers to your questions or get help for fixing your installation. This is a place to report bugs to developers, after your server has been debugged. You can find help debugging your system on our home user forums: https://help.nextcloud.com or, if you use Nextcloud in a large organization, ask our engineers on https://portal.nextcloud.com. See also  https://nextcloud.com/support for support options.

Nextcloud is an open source project backed by Nextcloud GmbH. Most of our volunteers are home users and thus primarily care about issues that affect home users. Our paid engineers prioritize issues of our customers. If you are neither a home user nor a customer, consider paying somebody to fix your issue, do it yourself or become a customer.

Guidelines for submitting issues:

* Please search the existing issues first, it's likely that your issue was already reported or even fixed.
    - Go to https://github.com/nextcloud and type any word in the top search/command bar. You probably see something like "We couldn’t find any repositories matching ..." then click "Issues" in the left navigation.
    - You can also filter by appending e. g. "state:open" to the search string.
    - More info on search syntax within github: https://help.github.com/articles/searching-issues
    
* This repository https://github.com/nextcloud/server/issues is *only* for issues within the Nextcloud Server code. This also includes the apps: files, encryption, external storage, sharing, deleted files, versions, LDAP, and WebDAV Auth
  
* SECURITY: Report any potential security bug to us via our HackerOne page (https://hackerone.com/nextcloud) following our security policy (https://nextcloud.com/security/) instead of filing an issue in our bug tracker.  

* The issues in other components should be reported in their respective repositories: You will find them in our GitHub Organization (https://github.com/nextcloud/)
  
* You can also use the Issue Template app to prefill most of the required information: https://apps.nextcloud.com/apps/issuetemplate
-->

<!--- Please keep this note for other contributors -->

### How to use GitHub

* Please use the 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to show that you are affected by the same issue.
* Please don't comment if you have no relevant information to add. It's just extra noise for everyone subscribed to this issue.
* Subscribe to receive notifications on status change and new comments. 


### Steps to reproduce
1. Share folder with group
2. Log in as LDAP user
3. Try to accept/reject shared folder

### Expected behaviour
The share should be accepted/rejected.

### Actual behaviour
Backend replies with 404: "Wrong share ID, share doesn't exist"

### Server configuration

**Operating system:**
Debian 10
**Web server:**
Nginx
**Database:**
Mariadb server 10.3
**PHP version:** 
7.3
**Nextcloud version:** 
19.0.1
**Updated from an older Nextcloud/ownCloud or fresh install:** 
Updated
**Where did you install Nextcloud from:** 
Zip archive download, version 14 think (done multiple upgrades since)
**Signing status:**
<details>
<summary>Signing status</summary>
No errors have been found.
</details>

**List of activated apps:**
<details>
<summary>App list</summary>
Enabled:
  - accessibility: 1.5.0
  - activity: 2.12.0
  - admin_audit: 1.9.0
  - apporder: 0.10.0
  - bruteforcesettings: 1.6.0
  - calendar: 2.0.3
  - cloud_federation_api: 1.2.0
  - comments: 1.9.0
  - contacts: 3.3.0
  - contactsinteraction: 1.0.0
  - dav: 1.15.0
  - deck: 1.0.5
  - event_update_notification: 1.0.2
  - external: 3.6.0
  - extract: 1.2.4
  - federatedfilesharing: 1.9.0
  - files: 1.14.0
  - files_external: 1.10.0
  - files_pdfviewer: 1.8.0
  - files_rightclick: 0.16.0
  - files_sharing: 1.11.0
  - files_trashbin: 1.9.0
  - files_versions: 1.12.0
  - files_videoplayer: 1.8.0
  - firstrunwizard: 2.8.0
  - groupfolders: 6.0.6
  - impersonate: 1.6.1
  - logreader: 2.4.0
  - lookup_server_connector: 1.7.0
  - meglalektura: 0.0.1
  - metadata: 0.12.0
  - music: 0.15.1
  - nextcloud_announcements: 1.8.0
  - notes: 3.6.0
  - notifications: 2.7.0
  - oauth2: 1.7.0
  - onlyoffice: 4.3.0
  - password_policy: 1.9.1
  - passwords: 2020.7.0
  - photos: 1.1.0
  - polls: 1.4.3
  - privacy: 1.3.0
  - provisioning_api: 1.9.0
  - recommendations: 0.7.0
  - serverinfo: 1.9.0
  - settings: 1.1.0
  - sharebymail: 1.9.0
  - survey_client: 1.7.0
  - systemtags: 1.9.0
  - talk_simple_poll: 1.1.1
  - tasks: 0.13.2
  - text: 3.0.1
  - theming: 1.10.0
  - theming_customcss: 1.6.0
  - twofactor_backupcodes: 1.8.0
  - updatenotification: 1.9.0
  - user_ldap: 1.9.0
  - viewer: 1.3.0
  - workflowengine: 2.1.0
Disabled:
  - encryption
  - federation
  - files_accesscontrol
  - files_automatedtagging
  - richdocuments
  - spreed
  - support
  - timetracker
  - workflow_script
</details>

**Nextcloud configuration:**
<details>
<summary>Config report</summary>
{
    "system": {
        "trusted_domains": [
            "megla.radiostudent.si"
        ],
        "dbtype": "mysql",
        "dbhost": "***REMOVED SENSITIVE VALUE***",
        "dbname": "***REMOVED SENSITIVE VALUE***",
        "dbuser": "***REMOVED SENSITIVE VALUE***",
        "dbpassword": "***REMOVED SENSITIVE VALUE***",
        "dbtableprefix": "oc_",
        "default_language": "sl",
        "default_locale": "sl_SI",
        "defaultapp": "files",
        "knowledgebaseenabled": true,
        "allow_user_to_change_display_name": true,
        "remember_login_cookie_lifetime": 1296000,
        "session_lifetime": 86400,
        "session_keepalive": true,
        "auth.bruteforce.protection.enabled": true,
        "skeletondirectory": "",
        "mail_domain": "***REMOVED SENSITIVE VALUE***",
        "mail_from_address": "***REMOVED SENSITIVE VALUE***",
        "mail_smtpdebug": false,
        "mail_smtpmode": "smtp",
        "mail_smtphost": "***REMOVED SENSITIVE VALUE***",
        "mail_smtpport": 25,
        "mail_smtptimeout": 10,
        "mail_smtpsecure": "",
        "mail_smtpauth": false,
        "mail_smtpauthtype": "LOGIN",
        "mail_smtpname": "***REMOVED SENSITIVE VALUE***",
        "mail_smtppassword": "***REMOVED SENSITIVE VALUE***",
        "mail_template_class": "\\OC\\Mail\\EMailTemplate",
        "mail_send_plaintext_only": false,
        "mail_sendmailmode": "smtp",
        "instanceid": "***REMOVED SENSITIVE VALUE***",
        "passwordsalt": "***REMOVED SENSITIVE VALUE***",
        "secret": "***REMOVED SENSITIVE VALUE***",
        "datadirectory": "***REMOVED SENSITIVE VALUE***",
        "version": "19.0.1.1",
        "overwrite.cli.url": "https:\/\/megla.radiostudent.si",
        "dbport": "",
        "installed": true,
        "ldapIgnoreNamingRules": false,
        "ldapProviderFactory": "OCA\\User_LDAP\\LDAPProviderFactory",
        "mysql.utf8mb4": true,
        "maintenance": false,
        "memcache.local": "\\OC\\Memcache\\APCu",
        "memcache.locking": "\\OC\\Memcache\\Redis",
        "redis": {
            "host": "***REMOVED SENSITIVE VALUE***",
            "port": 6379
        },
        "theme": "",
        "log_type": "file",
        "logfile": "\/var\/log\/nextcloud\/nextcloud.log",
        "logfilemode": 416,
        "loglevel": 1,
        "logdateformat": "F d, Y H:i:s",
        "app_install_overwrite": [
            "external",
            "timetracker"
        ],
        "simpleSignUpLink.shown": false,
        "updater.release.channel": "beta",
        "debug": false,
        "onlyoffice": {
            "verify_peer_off": true
        },
        "updater.secret": "***REMOVED SENSITIVE VALUE***"
    }
}
</details>

**Are you using external storage, if yes which one:** local/smb/sftp/...
yes, local
**Are you using encryption:** yes/no
no
**Are you using an external user-backend, if yes which one:** LDAP/ActiveDirectory/Webdav/...
yes, LDAP
#### LDAP configuration (delete this part if not used)
<details>
<summary>LDAP config</summary>
+-------------------------------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Configuration                 | s01                                                                                                                                                                                                                        |
+-------------------------------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| hasMemberOfFilterSupport      |                                                                                                                                                                                                                            |
| homeFolderNamingRule          |                                                                                                                                                                                                                            |
| lastJpegPhotoLookup           | 0                                                                                                                                                                                                                          |
| ldapAgentName                 |                                                                                                                                                                                                                            |
| ldapAgentPassword             | ***                                                                                                                                                                                                                        |
| ldapAttributesForGroupSearch  | cn                                                                                                                                                                                                                         |
| ldapAttributesForUserSearch   | displayName;gecos;mail                                                                                                                                                                                                     |
| ldapBackupHost                |                                                                                                                                                                                                                            |
| ldapBackupPort                |                                                                                                                                                                                                                            |
| ldapBase                      | dc=radiostudent,dc=si                                                                                                                                                                                                      |
| ldapBaseGroups                | ou=Groups,dc=radiostudent,dc=si                                                                                                                                                                                            |
| ldapBaseUsers                 | ou=Users,dc=radiostudent,dc=si                                                                                                                                                                                             |
| ldapCacheTTL                  | 600                                                                                                                                                                                                                        |
| ldapConfigurationActive       | 1                                                                                                                                                                                                                          |
| ldapDefaultPPolicyDN          |                                                                                                                                                                                                                            |
| ldapDynamicGroupMemberURL     |                                                                                                                                                                                                                            |
| ldapEmailAttribute            | mail                                                                                                                                                                                                                       |
| ldapExperiencedAdmin          | 1                                                                                                                                                                                                                          |
| ldapExpertUUIDGroupAttr       | entryUUID                                                                                                                                                                                                                  |
| ldapExpertUUIDUserAttr        | entryUUID                                                                                                                                                                                                                  |
| ldapExpertUsernameAttr        | uid                                                                                                                                                                                                                        |
| ldapExtStorageHomeAttribute   |                                                                                                                                                                                                                            |
| ldapGidNumber                 | gidNumber                                                                                                                                                                                                                  |
| ldapGroupDisplayName          | cn                                                                                                                                                                                                                         |
| ldapGroupFilter               | (|(objectclass=posixGroup)) |
| ldapGroupFilterGroups         | *listofgroups*                                                                                                       |
| ldapGroupFilterMode           | 0                                                                                                                                                                                                                          |
| ldapGroupFilterObjectclass    | posixGroup                                                                                                                                                                                                                 |
| ldapGroupMemberAssocAttr      | memberUid                                                                                                                                                                                                                  |
| ldapHost                      | **removed sensitive value**                                                                                                                                                                                                   |
| ldapIgnoreNamingRules         |                                                                                                                                                                                                                            |
| ldapLoginFilter               | (&(|(objectclass=inetOrgPerson))(|(uid=%uid)(|(mailPrimaryAddress=%uid)(mail=%uid))))                                                                                                                                      |
| ldapLoginFilterAttributes     |                                                                                                                                                                                                                            |
| ldapLoginFilterEmail          | 1                                                                                                                                                                                                                          |
| ldapLoginFilterMode           | 0                                                                                                                                                                                                                          |
| ldapLoginFilterUsername       | 1                                                                                                                                                                                                                          |
| ldapNestedGroups              | 0                                                                                                                                                                                                                          |
| ldapOverrideMainServer        |                                                                                                                                                                                                                            |
| ldapPagingSize                | 500                                                                                                                                                                                                                        |
| ldapPort                      | *ldapport*                                                                                                                                                                                                                        |
| ldapQuotaAttribute            |                                                                                                                                                                                                                            |
| ldapQuotaDefault              |                                                                                                                                                                                                                            |
| ldapTLS                       | 0                                                                                                                                                                                                                          |
| ldapUserAvatarRule            | default                                                                                                                                                                                                                    |
| ldapUserDisplayName           | displayname                                                                                                                                                                                                                |
| ldapUserDisplayName2          | mail                                                                                                                                                                                                                       |
| ldapUserFilter                | (|(objectclass=inetOrgPerson))                                                                                                                                                                                             |
| ldapUserFilterGroups          |                                                                                                                                                                                                                            |
| ldapUserFilterMode            | 0                                                                                                                                                                                                                          |
| ldapUserFilterObjectclass     | inetOrgPerson                                                                                                                                                                                                              |
| ldapUuidGroupAttribute        | auto                                                                                                                                                                                                                       |
| ldapUuidUserAttribute         | auto                                                                                                                                                                                                                       |
| turnOffCertCheck              | 0                                                                                                                                                                                                                          |
| turnOnPasswordChange          | 0                                                                                                                                                                                                                          |
| useMemberOfToDetectMembership | 1                                                                                                                                                                                                                          |
+-------------------------------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
</details>

### Client configuration
**Browser:**
Latest chrome/firefox
**Operating system:**
Debian linux
### Logs

<!--- Reports without logs might be closed as unqualified reports! -->

#### Web server error log
<details>
<summary>Web server error log</summary>
Checked, none applicable
</details>

#### Nextcloud log (data/nextcloud.log)
 <details>
<summary>Nextcloud log</summary>
Checked, none applicable
</details>

#### Browser log
<details>
<summary>Browser log</summary>
Network response:
<?xml version="1.0"?>
<ocs>
 <meta>
  <status>failure</status>
  <statuscode>404</statuscode>
  <message>Napačen ID mesta uporabe; mesto ne obstaja!</message>
 </meta>
 <data/>
</ocs>

</details>

**Comment by g1smo** [comment]
I did some digging in code and found out adding one line of code fixes the issue:
https://github.com/nextcloud/server/pull/21897

The problem was that when checking if the logged in user is a member of the group that the folder was shared with, the list of members was not populated.

**Comment by g1smo** [comment]
I've found some other issues and I'm investigating further; the fix I posted above is more of a workaround that seemed to take care of only some of the issues.

**Comment by g1smo** [comment]
Okay I updated to 19.0.2 and inquired further.
The bug was still there and the workaround above still helped.

After diving deeper into how LDAP group membership is checked, I've pinpointed the issue. It turns out that the method which checks group membership doesn't seem to work right on our openldap server configuration which uses memberUid for group membership.

I've commited a small fix, this time more granular, that seems to fix the issue:
https://github.com/nextcloud/server/pull/22530

I'd also like to note that this issue is present in some older, existing folder shares. I don't remember which version was used to make them, unfortunately.
